Nlp docs base & supported tasks (#1993)
deepchecks/deepchecks: docs/source/user-guide/nlp/supported_tasks.rst

> .. _supported_tasks:

Reviewer: labels have to be unique within a project...
```suggestion
.. _nlp_supported_tasks:
```

Reviewer: by the way - why did you add these all throughout? (just wondering cause I didn't see where they were used)

Author: I just copied the way the rest of the docs are built...

===================================

> Working with Labels and Predictions

Reviewer: question - do you think the title here would be clearer for the user if it's:
Supported Task Types

Or something along those lines? (e.g. we can add in brackets "passing labels and predictions" or such). and same for the Tabular user guide, to call it "Supported Models"

Author: I want to convey the message that this is the place to look to understand what formats we support. Same in tabular, that's why "Supported Models" doesn't sound great to me for tabular - it sounds like "we support XGBoost".

What about "Supported Tasks and Label Formats" or something along these lines?

===================================

Some checks, mainly the ones related to model evaluation, require labels and model predictions in order to run.
In the deepchecks nlp package, predictions are passed into the suite / check ``run`` method as pre-computed
predictions only (passing a fitted model is currently not supported).

> .. _supported_task__types:

Reviewer: ```suggestion
.. _nlp_supported_task__types:
```

Supported Task Types
====================

Deepchecks currently supports two NLP task types:

* **Text Classification**: Text classification is any NLP task in which a whole body of text (ranging from a sentence

>   to a document) is assigned a class (in the binary/multiclass case) or a certain set of classes (in the multilateral

Reviewer: if that's what you meant here

```suggestion
  to a document) is assigned a class (in the binary/multiclass case) or a certain set of classes (in the multilabel
```

  case). In both the binary, the multiclass and the multilabel case the class "belongs" / "classifies" the whole text

>   sample. Examples for such tasks are Sentiment Analysis, Topic Extraction, detecting harmful content and more.

Reviewer: ```suggestion
  sample. Examples for such tasks are: Sentiment Analysis, Topic Extraction, harmful content detection and more.
```

Reviewer: not sure if with or without capital, but the mix didn't look good IMO. feel free to reject...

Author: Think without capital because I made this up rather than this being an accepted term (as far as I know)

* **Token Classification**: Token Classification is any NLP task in which each word (or to be more accurate, token) in

>   the text sample is assigned a class of it's own. In many cases most tokens will belong to a "background" class,

Reviewer: ```suggestion
  the text sample is assigned a class of its own. In many cases most tokens will belong to a "background" class,
```

>   allowing the model to focus on the interesting tokens. Examples for such tasks are Named Entity Recognition, Part
>   of Speach annotation (in which all tokens have a non-background class).

Reviewer: ```suggestion
  allowing the model to focus on the interesting tokens. Examples for such tasks are: Named Entity Recognition, Part-of-speech annotation (in which all tokens have a non-background class).
```

Reviewer: feel free to reject (but if so fix the typo in speech)

> .. _supported_labels__predictions_format:

Reviewer: ```suggestion
.. _nlp_supported_labels__predictions_format:
```

> Supported Labels and Predictions Format

Reviewer: To be honest when I read this section - it makes me feel that it's more convenient for the hierarchy to be:

Task Type 1
===========
Label Format
--------------

Prediction Format
---------------------

Task Type 2
==========
Label Format
--------------

Prediction Format
-------------------

and not the current (label, prediction as outer, task type as inner), as this is more natural in the user's flow.... WDYT?
If so - we can open issues (for tabular, nlp, etc...) and tag them as "first good issues".
Anyway wouldn't consider it as urgent, but could be helpful....

Author: Done

=======================================

> While labels are passed while constructing the :class:`TextData <text_data.TextData>` object, predictions are passed

Reviewer: the link to class doesn't work... maybe full path required?

> While labels are passed while constructing the :class:`TextData <text_data.TextData>` object, predictions are passed

Reviewer: ```suggestion
While labels are passed when constructing the :class:`TextData <text_data.TextData>` object, predictions are passed
```

separately to the ``run`` method of the check / suite. Labels and predictions must be in the format detailed in this

> section, according the task type.

Reviewer: ```suggestion
section, according to the task type.
```

> Label Formats:

Reviewer: ```suggestion
Label Formats
```

--------------

> * **Text Classification** Label - For text classification the accepted label format differs between multilabel and

Reviewer: I suggest to have the task type as an additional sub-title hierarchy...
(github markdown parses this incorrectly, so instead of text added picture)

<img width="220" alt="CleanShot 2022-09-12 at 15 37 12@2x" src="https://user-images.githubusercontent.com/33841818/189655303-88f53845-a87a-46e8-8845-8a2eba4dd1f3.png">

  single label cases. For single label data, the label should be passed as a sequence of labels, with one entry
  per sample that can be either a string or an integer. For multilabel data, the label should be passed as a
  sequence of sequences, with the sequence for each sample being a binary vector, representing the presence of
  the i-th label in that sample.
* **Token Classification** Label - For token classification the accepted label format is a sequence of sequences,
  with the inner sequence containing tuples in the following format: (class_name, span_start, span_end).
  span_start and span_end are the start and end character indices of the token within the text, as it was

>   passed to the raw_text argument. Each upper level sequence contains a sequence of tokens for each sample.

Reviewer: ```suggestion
  passed to the raw_text argument. Each outer sequence contains the sequence of tokens for each sample.
```

This sentence wasn't clear to me.. tried making it clearer (assuming that's what you meant)
Consider dropping completely this part, might make it clearer...

>>> text_classification_label_multiclass = [0, 0, 1, 2]
>>> text_classification_label_multilabel = [[0, 0, 1], [0, 1, 1], [1, 0, 1], [0, 0, 0]]
>>> token_classification_label = [[('class_1', 0, 2), ('class_2', 7, 10)], [('class_2', 42, 54)], []]

> Prediction Formats:

Reviewer: ```suggestion
Prediction Formats
```

-------------------

The accepted formats for providing model predictions and probabilities are detailed below

**Text Classification**

> *Single Class Predictions*

Reviewer: and have these in bold instead of italic. or maybe better - a further subtitle...
<img width="210" alt="CleanShot 2022-09-12 at 15 39 33@2x" src="https://user-images.githubusercontent.com/33841818/189655795-11774a81-4610-4b26-82ac-4fa24d165f57.png">

> * predictions - A sequence of class names or indices with one entry per sample, matching the set of classes

Reviewer: Then can make here all of the following "predictions", "probabilities" texts bold to stand out

  present in the labels.
* probabilities - A sequence of sequences with each element containing the vector of class probabilities for

>   each sample. Each such vector should have one probability per class according to the class (sorted) order, and

Reviewer: not sure - what is the "sorted" order? (who sorts this? confusing a bit, is there anything I should actively do to verify consistency with their "order" somewhere else? or just have a consistent order with myself? and are the class names displayed somewhere? etc.)

  the probabilities should sum to 1 for each sample.

>>> predictions = ['class_1', 'class_1', 'class_2']

> >>> probabilities = [[0.2, 0.8], [0.5, 0.5], [0.3, 0.7]]

Reviewer: I think somewhere it's worth stating explicitly the binary use case - that also in that case it's still a probability per class.
It's trivial from the example, and yet can still be easy to miss deducing.

suggestion: to add a short comment above (in code, write above the code example) relating to that

*Multilabel Predictions*

* predictions - A sequence of sequences with each element containing a binary vector denoting the presence of
  the i-th class for the given sample. Each such vector should have one binary indicator per class according to
  the class (sorted) order. More than one class can be present for each sample.
* probabilities - A sequence of sequences with each element containing the vector of class probabilities for
  each sample. Each such vector should have one probability per class according to the class (sorted) order, and
  the probabilities should range from 0 to 1 for each sample, but are not required to sum to 1.

>>> predictions = [[0, 0, 1], [0, 1, 1]]
>>> probabilities = [[0.2, 0.3, 0.8], [0.4, 0.9, 0.6]]

**Token Classification**

* predictions - A sequence of sequences, with the inner sequence containing tuples in the following
  format: (class_name, span_start, span_end, class_probability). span_start and span_end are the start and end
  character indices  of the token within the text, as it was passed to the raw_text argument. Each upper level

>   sequence contains a sequence of tokens for each sample.

Reviewer: same note here about the "upper level" (whichever wording you choose to go with)
